### PR TITLE
Feature Bundle

### PR DESCRIPTION
**Pull request by PhRX**
This PR supersedes #173, which pretty much is obsoloete given the volume of changes.

It almost concludes the work I wanted to do. An overview of the changes can be found below.

Since the amount of code is probably larger again than what's in #173, reviewing isn't getting any easier. I'm aware of this. but IMHO having to wait for reviews should not hold up progressing on the project if the progress is on a large enough scale.

I'd like to propose the following procedure :

* Create a new branch in the original HP repository called "approval"
* Close the 2 PRs (this one and #173)
* Let me create a new PR pointing at the "approval" branch
* Merge the PR into "approval"
* In the instructions, make people aware there's an alternative approval branch which has more features
* Let people use it for a while; if possible, also review the code
* If possible, have somewhere where people can vote for/against the "approval" changes
* If enough people agree that it works and is a worthwhile evolution, and any bugs or issues found by users have been fixed, integrate the branch into master

This way, we add an alternative standard for PR acceptance next to code review : enough users being happy as yardstick. Also, a branch on master is more visible than "just another fork".

Please let me know whether that works for y'all.

So, in addition to what's in #173, the following features have been added : 

* Fix for #36 : the aggregation options (see below) allow collapsing thread trees into an aggregated tree
* Fix for #38 : the aggregation options allow inclusion of the line nr, which disambiguates the method calls (but having the signature somehow would be even nicer)
* Fix for #78 : during "live" profiling, the profile can be "frozen", so the front-end no longer receives updates. When unfreezing, the view will be updated to the current situation.
* Fix for #175 : large log files can now be opened. The log files are no longer memory-mapped in their entirety, instead the buffer now is a "moving window".
* Fix for #179 : top-level frames (children of thread-level) are now aggregated correctly
* The Self Time and Total Time, available in the agent output, is now exposed in all views
* Aggregation can now be switched as follows : 
  - Frames : By Fully Qualified Method Name, By FQMN + Line Nr, By FQMN + BCI
  - Threads : All threads together, Threads by Id (tree views only), Threads by name (tree views only)
* The source profile (LeanProfile) now contains the minimum of information without redundancy, to ensure the collection thread has to do as little work as possible
* Pull mechanism for profiles : the front-end now requests updates, which allows it to control the update rate (but no UI controls have yet been provided except for the "freeze" buttons)
* Aggregation happens lazily and is cached
* Tree Diff view for comparing two profiles
* Added Infobar with help information
* In the Flat view (list of methods), when selecting a method, its ancestors (methods directly or indirectly calling the selected method) and descendants (methods directly and indirectly called by the selected method) are shown in two trees at the bottom of the screen
* In the Tree view, when selecting a method, its descendants are aggregated into a list shown at the bottom
* Added more columns in all views, with homogenous labeling
* Added QuickFilter function for filtering the key
* Prevent opening the same source twice
* Fixed VirtualMachine source issue where the log file produced by an agent couldn't always be located
* Included font with non-proportional digit characters for more readable number tables
* Smaller default font size (through CSS)
* On refresh, expand TreeViews by 1 level
* Sorting is preserved (at least in flat views) when the aggregation is updated
* I18N for all text (except some enum choices, still got to figure out the pattern for that)
* Remember parent directory when opening a log file, to prevent lots of renavigating to the same or similar location

At code level, the following was changed :

* New core.aggregation package added, containing :
  - aggregator package with the actual aggregation transformation logic
  - filter package with more generalized filtering logic
  - grouping package with "key grouping" logic for aggregation
  - result package with the aggregation data structures
* New core.collector.lean package containing the new LeanLogCollector which implements the pull mechanism and which produces the more lightweight "LeanProfile"s
* New core.profiles.lean package containing the data structures for the new "lean" profiles
* Extracted common functionality from controllers and put it into abstract superclasses for code reuse
* Use JavaFX tasks to decouple profile generation and initial aggregation from front-end
* Documented many methods and classes (but some documentation eneds to be revisited due to recent fixes and refactorings)

Still to do, optionally :

* Percentages are now calculated relative to the "global aggregation", over threads. Allow reference switching to thread-level where appropriate (i.e. in tree views aggregated per thread, show percentages relative to thread total rather than relative to sum over all threads)
* In addition, maybe add option not to take the "error" counts into account when aggregating ?
* Highlight filter hits (useful in tree filtering)
* Maybe add aggregated ancestors/descendants views to flat diff and tree diff ?
* Make tabs/profiles closeable
* Add Preferences, including styling options (font size, ...)
* Clearer QuickFilter UI ? (show currently active quickfilter, to avoid confusion when something was typed in the text field but not "activated")
* Add Clear QuickFilter button / Clear All Filters button
* Remember selection when aggregation is updated
* Remember which tree nodes were expanded/collapsed when aggregation is updated

### CODE REVIEW

**Comment by RichardWarburton** [comment]
Hi @PhRX ,

Thanks for continuing to work on this stuff. I don't really like the "approval" branch idea, sounds like it'll cause a lot more confusion. Better to just progress with the review. Since this PR is more advanced than the last one then I'm happy to have a go at reviewing this one. I propose we close #173 and focus on this one. What do you think?

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/info/NumericInfo.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+ */
+public class NumericInfo
+{
+    // Instance Properties
+
+    private BigInteger selfTime;
+    private BigInteger totalTime;
+
+    private int selfCnt;
+    private int totalCnt;
```
Comment (chhsiao90) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/info/NumericInfo.java]: Is that selfCnt or totalCnt had the risk for overflow?
int -> BigInteger will be better?

Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/info/NumericInfo.java]: Given that this is a number of samples, and that INT_MAX is rather large, it would take a program which samples 1000 times a second nearly 50 days of profiling to overflow. I would not worry.

File: src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java [hunk 2]
```diff
@@ -63,6 +66,19 @@ public boolean isAgentLoaded()
...
+        int agentPathStart = vmArgs.indexOf("-agentpath") + "-agentpath".length();
+        int agentPathEnd = vmArgs.indexOf(' ', agentPathStart);
+        String agentPath = vmArgs.substring(agentPathStart, agentPathEnd);
+        int logPathStart = agentPath.indexOf("logPath=") + "logPath=".length();
+        int commaPos = agentPath.indexOf(",", logPathStart);
+        int spacePos = agentPath.indexOf(' ', logPathStart);
+        int logPathEnd = commaPos > 0 ? commaPos : (spacePos > 0 ? spacePos : agentPath.length());
+
+        return new FileLogSource(new File(agentPath.substring(logPathStart, logPathEnd)));
+    }
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java]: Are these problems not present with previous version?

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java [hunk 3]
```diff
@@ -40,10 +40,25 @@ public Method(long methodId, String fileName, String className, String methodNam
...
+                   String className,
+                   String methodName,
+                   boolean dummy)
+    {
+        this.methodId = methodId;
+        this.fileName = fileName;
+        this.className = className;
+        this.methodName = methodName;
+    }
+
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java]: out of scope, lets not add further changes to this already massive PR.

File: src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java (hunk 2 above)
Comment (chhsiao90) [inline src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java]: No,
previous version seems only used [getLogSource()](https://github.com/RichardWarburton/honest-profiler/blob/master/src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java#L66-L82) to get latest log file at userDir
And current version used [getLogSourceFromVmArgs()](https://github.com/PhRX/honest-profiler-fx-ui/blob/master/src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java#L69-L80) depends on vmArgs

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/ThreadMeta.java [hunk 4]
```diff
@@ -67,12 +67,27 @@ public void accept(LogEventListener listener)
...
 
+    public ThreadMeta update(ThreadMeta newMeta)
+    {
+        if (newMeta.threadName != null && !newMeta.threadName.isEmpty())
+        {
+            threadName = newMeta.threadName;
+        }
+
+        return this;
+    }
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/ThreadMeta.java]: Fixed in upcoming commit, along with ThreadInfo which follows the same pattern.

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        data = new NumericInfo(nanos);
+        this.parent = parent;
+        childMap = new HashMap<>();
+    }
+
+    /**
+     * Copy constructor.
+     *
+     * @param source the source LeanNode which is being copied
+     * @param new Parent the parent of the copy (which itself generally is a copy)
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java]: Fixed in upcoming commit.

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+     */
+    public boolean isThreadNode()
+    {
+        return false;
+    }
+
+    // Aggregation Methods
+
+    /**
+     * Aggregates teh information from a child {@link StackFrame} into the children of this LeanNode, updating the total
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java]: Fixed in upcoming commit.

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    /**
+     * Copy method for children.
+     *
+     * @param newParent the new parent for the children
+     * @return a copy of this object
+     */
+    public LeanNode copy(LeanNode newParent)
+    {
+        return new LeanNode(this, newParent);
+    }
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java]: Fixed in upcoming commit, by getting rid of the two copy() methods. One was unused, and the one referenced here was used only in a single location where it didn't hurt readability to replace it by a constructor call.

File: src/main/java/com/insightfullogic/honest_profiler/ports/javafx/controller/ProfileDiffRootController.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        controllerMap.put(FLAT, asList(flatController));
+        controllerMap.put(TREE, asList(treeController));
+    }
+
+    // Instance Accessors
+
+    /**
+     * Sets the {@link ApplicationContext} and propagates it to any contained controllers.
+     *
+     * @param appCtx the {@link ApplicationContext}
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/ports/javafx/controller/ProfileDiffRootController.java]: Fixed in upcoming commit.

File: src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java (hunk 2 above)
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java]: Hm, my bad the. I'll check and fix. The intention was to revert to the original logic if no logpath parameter was found, so as to only extend the number of cases in which the mechanism works correctly.

Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java]: Fixed in upcoming commit. I hardened the code, by introducing early checks if what we're looking for is present at all, and revert to original getLogSource() if it isn't. Also, the final File is checked for existence, and if it doesn't, revert to the original getLogSource() as well.

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/info/NumericInfo.java (hunk 1 above)
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/info/NumericInfo.java]: Thanks @nitsanw for doing the calculations for me :) This is pretty much what I intuited, when deciding between long and int (BigInteger would have been overkill), but I never actually did the math. The only reason for the BigInt is that when working with nanos, you potentially do get really big numbers.

**Comment by PhRX** [comment]
Thanks @chhsiao90 and @nitsanw for the review.

@RichardWarburton I forgot to reply, but I agree (and I already closed the old PR), although I'd like to do something about the visibility of the new stuff - a fork is in my (limited) experience very easily ignored entirely.

I'll finish up the documentation - only some utility classes are undocumented, and I'll probably write some package documentation which should make it clearer how the concepts work together - I can imagine that things like Grouping and Target may be a bit confusing or obscure when first coming across them. Some pictures or at least some text with a top-down explanation might clarify (comments in the code are by nature more bottom-up).

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+     * data for this LeanNode in the process.
+     *
+     * The add method is called from the LogCollector, on the parent with the next (potentially last) child from the
+     * trace.
+     *
+     * @param nanos the number of nanoseconds between the {@link TraceStart} preceding the stack trace sample and the
+     *            {@link TraceStart} following it
+     * @param child the child {@link FrameInfo} of the current LeanNode
+     * @param last a boolean indicating if the child is the last in the stack trace sample
+     * @return this node
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java]: Fixed in upcoming commit.

File: src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+                new LeanNode(child, nanos, this)
+                // New child is not the last in stack, use "Non-self constructor".
+                : new LeanNode(child, this))
+                // Aggregate the information into an existing child
+                : (last ?
+                // Child is last in stack, use the "self add" on existing child
+                    v.addSelf(nanos) :
+                    // Child is not last in stack, return existing child (its "total" numbers will be updated when its
+                    // child is added)
+                    v));
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java]: Yes, excellent catch ! This is indeed a lot simpler. As a bonus I can get rid of the "self constructor" as well, as well as a now obsolete constructor in NumericInfo.

Thanks !
Fixed in upcoming commit.

File: src/main/java/com/insightfullogic/honest_profiler/core/collector/lean/LeanLogCollector.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+ * A {@link LeanProfile} will only be emitted when requested, or when the end of the log file is reached. When the final
+ * {@link LeanProfile} has been emitted after a log file has been processed, requests no longer have any effect.
+ *
+ * As long as no stacks have been received, nothing will be emitted.
+ */
+public class LeanLogCollector implements LogEventListener, ProfileSource
+{
+    // Class Properties
+
+    private static final long SECONDSTONANOS = 1000 * 1000 * 1000;
```
Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/lean/LeanLogCollector.java]: Fixed in upcoming commit.

**Comment by PhRX** [comment]
To all potential reviewers : in my most recent [commit](https://github.com/RichardWarburton/honest-profiler/pull/184/commits/c885722c6b184ed74fde0d2687f412ec0c3d6ae4) I added the package description for the core.profiles.lean package.

I **very very highly recommend reviewing it early on**, since I tried to describe in detail the profiler basics, the internal structure of the LeanProfile data structure (which the core data structure on top of which all aggregations are built), and how the numerical data in there is (minimally) aggregated.

As a result, the text pretty accurately reflects many of my presumptions and assumptions when I wrote the code. Therefore, **if there are any conceptual mistakes in this text**, it is fairly certain that **those mistakes are reflected in the code** !

Unfortunately for me the converse is not true - if the text is accurate and sound, it doesn't mean the actual code is correct :-/

Of course an additional benefit of reading it is that the LeanProfile structure should be clearer afterwards.

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java (hunk 3 above)
Comment (RichardWarburton) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java]: Thanks for helping with review this PR @nitsanw and @chhsiao90 - greatly appreciated! I still think we should keep to one build system and not have it as part of this PR. I don't mind if we move to gradle in future, but let's not bundle that up in these changes.

**Comment by RichardWarburton** [comment]
Hi,

I've tried out the new JavaFX UI and in general it looks good. I like the multiple profile tabs and the comparison view - good improvements.

I've noticed a couple of things that look like regressions:

 * If you start another Java program, it will appear in the monitor menu, but if you quit the other program then its entry won't disappear from the menu.
 * If you file->quit or click X in the ui after you have opened a log it closes the window, but doesn't actually exit the Java process.
 * When I try to monitor a program I get the following exception:

```
java.lang.StringIndexOutOfBoundsException: String index out of range: -11
	at java.lang.String.substring(String.java:1967)
	at com.insightfullogic.honest_profiler.core.sources.VirtualMachine.getLogSourceFromVmArgs(VirtualMachine.java:73)
	at com.insightfullogic.honest_profiler.ports.javafx.model.task.InitializeProfileTask.monitor(InitializeProfileTask.java:80)
	at com.insightfullogic.honest_profiler.ports.javafx.model.task.InitializeProfileTask.call(InitializeProfileTask.java:45)
	at com.insightfullogic.honest_profiler.ports.javafx.model.task.InitializeProfileTask.call(InitializeProfileTask.java:22)
	at javafx.concurrent.Task$TaskCallable.call(Task.java:1423)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by PhRX** [comment]
Hi @RichardWarburton,

Thanks for the testing ! I addressed the 3 issues you mentioned :

- The JVM Stop not being detected seems not to have been a regression, but an issue in the "original" code : it would always try to attach a VM also in the "close" case. "attach" is now called only once as well (which addresses the TODO comment in there, which I now realize I forgot to remove). Not sure if there is a case for "remembering" descriptors after a failed attach, and trying to attach again on subsequent polls ?

- The Java process would exit eventually, but I made it snappy by ensuring the executor service in the application context gets shut down on application close, and that LocalMachineSource doesn't ignore being interrupted but triggers the ThreadedAgent interrupt logic properly.

- Are you sure that you're running the latest version from the integration thread ? The line numbers in the exception didn't match up with mine, and in https://github.com/RichardWarburton/honest-profiler/pull/184/commits/54f79108f2bf2aec442889ce0fa1bf5e13171e7b I hardened the "logPath extraction" code quite a bit (precisely to avoid the OutOfBounds exceptions like you're seeing). I reviewed the code again, and added a minor fix, but that corner case (the args containing "logPath" but not "logPath=") should not have applied to your test.

**Comment by PhRX** [comment]
@RichardWarburton @chhsiao90 I was very focused on finishing the documentation yesterday, I didn't have time yet to remove the gradle bits (again). Will do so shortly.

@chhsiao90 and others : Is the landing page important ? Which functions or information would you like to see there ? It's not very difficult to reinstate but its functions have effectively been superseded by the menu options, and at the moment, once a profile has been opened, the landing page would never be seen again (because I haven't yet implemented any profile closing functions). That said, while mulling this yesterday I did realize the initial view looks a bit early 90s :-) (i.e. unwelcoming big blank area of nothing).

Suggestions and feedback more than welcome !

**Comment by chhsiao90** [comment]
@PhRX - landing page is not very necessary and seems not import with this PR.
I think keep empty in current version is fine :)

**Comment by RichardWarburton** [comment]
Hi @PhRX re: "Are you sure that you're running the latest version from the integration thread?" - I cloned a repo from your PR and used the compiled binary from that to produce the log. It was the latest commit as of the comment. I'll re-test this evening with your latest pushes into this PR.

**Comment by PhRX** [comment]
Hi @RichardWarburton  I didn't even know you could clone from a PR. If you want me to, I can update the code to collect more information and make the Exception describe precisely what went wrong. I just re-eyeballed the current state and I don't see anything anymore that could cause trouble - either the agent was attached to the JVM with -agentPath + a logPath arg, in which case the file described by logPath will be returned, otherwise the old code is invoked.

Thanks for testing !

**Comment by RichardWarburton** [comment]
After both using the build of this for a while and talking to @nitsanw I've decided the best thing is to merge this PR and cleanup any remaining issues afterwards. But please, please, please send single-purpose PRs in future!
